### ISSUE TITLE
Avoid tight coupling

### ISSUE THREAD

**Issue by mcollina** [feugy/mini-service#4]
Let's consider process A and process B. They share different codebases. However, it seems that for issuing a call from A to B using mini-service, A needs to have access to B source code.
This creates a tight coupling between the two processes.

It would be awesome if this could be avoided, maybe by using an external definition, like swagger (maybe with https://github.com/glennjones/hapi-swagger). In this way we might pull the definition of the B into A source code, and not its implementation.

**Comment by feugy** [comment]
Hello Mateo !

Given that A is any kind of nodejs application, and B is an package using mini-service, it's true that A needs to depend on B (as an NPM dependency).

The original idea behind mini-service is to allow to run B either remotely (in a dedicated nodejs µService accessible through Http), either locally (in A's nodejs process, as a monolith).

You can start to build your monolith, and then split it in different µServices after a while.

I agree this implies a tight coupling, but to allow monolithic usage, I don't see how we can avoid giving A access to B's code source.

That is said, I'll try to search for an alternative where B owner might decide to publish his package for monolith usage (A will use B as a dependency), or for µService usage (A will depend on B's definition only).
By doing that, B might publish two flavors of his package: one for monolith usage, and one for µService deployment.

**Comment by mcollina** [comment]
Ideally the client can be generic, and just load some form of spec in JSON format:

``` js
const miniService = require('mini-service')
const client = miniService.createClient(require('./servic-definition.json'), 'http://my.service')

client.doSomething().then()
```

**Comment by feugy** [comment]
When `B` is deployed remotely, it already exposes [a kind of descriptor](https://github.com/feugy/mini-service/blob/4ccc6d867f1a2d3bb0c0d5bfbee290c9aa2da4b9/lib/server.js#L44).

So it's also doable to have a dedicated `mini-service-client` package that would only include `lib/client.js` file.

This way, `A` can only depend on `mini-service-client` in case of a remote `B`.
And in case of a monolith, I'll find a easy way to configure it so it loads `B`'s APIs locally.

**Comment by mcollina** [comment]
The point is not having that package. Having to maintain a client for each microservice is an explosion in maintenance costs. You want to have that definition pulled in inside the application code, and use a stock client that digest that.

**Comment by feugy** [comment]
My explanation was probably bad: `mini-service-client` can be a generic client.
We're on the same line 😄

**Comment by feugy** [comment]
PR #7 ready !

**Comment by feugy** [comment]
Addressed in version 2.0.0.
Mini-client is a generic client that doesn't force the caller code to have access to service code.
